src/documentation/io/testing.md

```diff
@@ -108,11 +108,464 @@ This allows the I/O transform's unit tests to act as if they have a perfect netw
 
+
+## I/O Transform Integration Tests {#i-o-transform-integration-tests}
+
+> We do not currently have examples of Python I/O integration tests or integration tests for unbounded or eventually consistent data stores. We would welcome contributions in these areas - please contact the Beam dev@ mailing list for more information.
+
+### Goals  {#it-goals}
+
+*   Allow end to end testing of interactions between data stores, I/O transforms, and runners, simulating real world conditions.
+*   Allow both small scale and large scale testing.
+*   Self contained: require the least possible initial setup or existing outside state, besides the existence of a data store that the test can modify.
+*   Anyone can run the same set of I/O transform integration tests that Beam runs on its continuous integration servers.
+
+
+### Integration tests, data stores, and Kubernetes {#integration-tests-data-stores-and-kubernetes}
+
+In order to test I/O transforms in real world conditions, you must connect to a data store instance.
+
+The Beam community hosts the data stores used for integration tests in Kubernetes. In order for an integration test to be run in Beam's continuous integration environment, it must have Kubernetes scripts that set up an instance of the data store.
+
+However, when working locally, there is no requirement to use Kubernetes. All of the test infrastructure allows you to pass in connection info, so developers can use their preferred hosting infrastructure for local development.
+
+
+### Running integration tests {#running-integration-tests}
+
+The high level steps for running an integration test are:
+1.  Set up the data store corresponding to the test being run
+1.  Run the test, passing it connection info from the just created data store
+1.  Clean up the data store
+
+Since setting up data stores and running the tests involves a number of steps, and we wish to time these tests when running performance benchmarks, we use PerfKit Benchmarker to manage the process end to end. With a single command, you can go from an empty Kubernetes cluster to a running integration test.
+
+However, **PerfKit Benchmarker is not required for running integration tests**. Therefore, we have listed the steps for both using PerfKit Benchmarker, and manually running the tests below.
+
+
+#### Using PerfKit Benchmarker {#using-perfkit-benchmarker}
+
+Prerequisites:
+1.  [Install PerfKit Benchmarker](https://github.com/GoogleCloudPlatform/PerfKitBenchmarker)
+1.  Have a running Kubernetes cluster you can connect to locally using kubectl
+
+You won't need to invoke PerfKit Benchmarker directly. Run mvn verify in the directory of the I/O module you'd like to test, with the parameter io-it-suite.
+
+Example run with the direct runner:
+```
+mvn verify -Dio-it-suite -Dio-it-suite-local -pl sdks/java/io/jdbc,sdks/java/io/jdbc -DpkbLocation="/Users/me/dev/PerfKitBenchmarker/pkb.py" -DforceDirectRunner -DintegrationTestPipelineOptions=["--myTestParam=val"]
```

Comment:
Not for this version of the document but maybe for the future we have to maybe unify this -DforceDirectRunner to come from the pipelineoptions or at least be explicitly defined like -Drunner=DataflowRunner (even if it is repeated at least it would be future proof)